src/_pytest/capture.py

```diff
@@ -686,16 +704,13 @@ def snap(self):
 
 
 class TeeSysCapture(SysCapture):
-    def __init__(self, fd, tmpfile=None):
-        name = patchsysdict[fd]
-        self._old = getattr(sys, name)
-        self.name = name
-        if tmpfile is None:
-            if name == "stdin":
-                tmpfile = DontReadFromInput()
-            else:
-                tmpfile = CaptureAndPassthroughIO(self._old)
-        self.tmpfile = tmpfile
+    def __init__(self, fd: int) -> None:
```

Comment:
I'm not sure this is really better, still duplicates some logic, and now it relies on `SysCaptureBinary` implementation instead of being independent.

Two possible alternatives, both are not ideal too but a bit better IMO:

1. Add a `passthrough: bool` argument to `SysCaptureBinary` and use `PassthroughCaptureIO` if true.
2. Add class attribute `CaptureIOClass: Type[CaptureIO] = CaptureIO` to `SysCaptureBinary` and use that in `__init__`, and override to `PassthroughCaptureIO` in `TeeSysCapture`.

Reply:
Well, I think it is quite simple/good.
Keep in mind that the "old" is only used for an assert, and can be removed.
As for your 2nd suggestion: this is basically what passing in `tmpfile` does, isn't it?

Reply:
> Keep in mind that the "old" is only used for an assert, and can be removed.

You mean `_old` right?

>  As for your 2nd suggestion: this is basically what passing in tmpfile does, isn't it?

\<offtopic>

The `tmpfile` argument exists for by `FDCaptureBinary`. I was surprised initially to see that `FDCaptureBinary` uses `SysCapture` under the hood. Turns out `FDCaptureBinary` not only redirects the FDs to a tempfile, it also *additionally* patches the `sys` file objects to redirect to the tempfile (this is where the argument is needed). It seemed redundant to me -- if we already redirect the FDs, there is no need to patch `sys` since the data will reach the tempfile anyway. So I tried to remove the `sys` patching from `FDCaptureBinary`. It actually works fine, but the problem is the flushing - if we don't patch `sys`, then data that remains in the file object buffer doesn't end up captured, i.e. an explicit `flush()` is sometimes needed before the capture fixture can be examined.

If backward compatibility wasn't a concern, I think that this is actually the better behavior for `capfd`. The implicit `flush()`s prevent a test from being able to verify that data is actually present in the underlying file at a point of time.

But, we can't break it now... So maybe it makes sense to add a new "`fdonly`" mode, and make `fd` an alias for "`fd+sys`" mode (for clarity).

\</offtopic>

TL;DR: I think the `tmpfile` argument should go away eventually, so wouldn't want to add new uses for it...